torch/csrc/jit/passes/bailout_graph.cpp

```diff
@@ -0,0 +1,216 @@
+#include <torch/csrc/jit/passes/alias_analysis.h>
+#include <torch/csrc/jit/passes/bailout_graph.h>
+#include <torch/csrc/jit/passes/constant_pooling.h>
+#include <memory>
+#include <unordered_set>
+
+namespace torch {
+namespace jit {
+
+struct BailOutGraphBuilderForNode {
+  BailOutGraphBuilderForNode(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {
+    copy_graph_ = std::make_shared<Graph>();
+  }
+
+  // capture `old_value` into the bailout graph
+  // by creating a new input and mapping
+  // `old_value` to it
+  Value* addNewInputForValue(Value* old_value) {
+    auto node = old_value->node();
+    // this reduces the number of inputs to a bailout graph significantly
+    // making it easier to debug
+    if (node->kind() == prim::Constant) {
+      auto new_const = copy_graph_->createClone(node, {nullptr});
+      copy_graph_->block()->appendNode(new_const);
+      return new_const->output();
+    }
+
+    live_inputs_.push_back(old_value);
+    auto new_value = copy_graph_->block()->addInput();
+    return mapExistingInputForValue(old_value, new_value);
+  }
+
+  Value* mapExistingInputForValue(Value* old_value, Value* new_value) {
+    this->old_to_new_[old_value] = new_value;
+    new_value->copyMetadata(old_value);
+    return new_value;
+  }
+
+  Value* getOrAddInputForValue(Value* v) {
+    if (this->old_to_new_.count(v) == 0) {
+      return addNewInputForValue(v);
+    } else {
+      return this->old_to_new_[v];
+    }
+  }
+
+  // buildBailOutBlockFrom builds a bailout graph from
+  // a given node `n` until the end of the owning block
+  // If `n` belongs to `prim::If` or `prim::Loop`
+  // buildBailOutGraphForOwningNode continues
+  // from block's owning node (e.g. `prim::If` or
+  // `prim::Loop`)
+  void buildBailOutBlockFrom(Node* n) {
+    auto* block = copy_graph_->block();
+    auto b = n->owningBlock();
+    graph_node_list_iterator it(n, kNextDirection);
+    for (; it != b->nodes().end(); it++) {
+      auto env = [this](Value* v) { return getOrAddInputForValue(v); };
+
+      auto node = *it;
+
+      auto new_node = block->appendNode(copy_graph_->createClone(node, env));
+      for (size_t i = 0; i < node->outputs().size(); ++i) {
+        auto oo = node->outputs()[i];
+        auto no = new_node->outputs()[i];
+        old_to_new_[oo] = no;
+        no->copyMetadata(oo);
```

Comment:
I am pretty sure this is already called by createClone. We should not call it again, because then people cargo cult it into new places thinking it is needed because they see it in other code.